### PR TITLE
feat(TDP-3333): read only preparation

### PR DESCRIPTION
**Pull request by charlesnguyen**
**Link to the JIRA issue**
https://jira.talendforge.org/browse/TDP-3333

**Please check if the PR fulfills these requirements**
- [x] The commit(s) message(s) follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md#commit-message-format)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [x] The code coverage on new code is > 75 % for backend and > 95% for frontend
- [x] The new code does not introduce new technical issues (sonar / eslint)
- [x] Functional tests have been performed

**Please check the browsers you've tested on**
- [x] Chrome, Firefox, Safari, Edge, IE11
- [ ] No, that's bad, this PR should not be merged !
- [ ] No, and no need to (backend changes only)

**(Optional) What is the current behavior?**
(Additional information to the Jira)

**(Optional) What is the new behavior?**
(Additional information to the Jira)

**(Optional) Other information**:

### CODE REVIEW

**Comment by build-talend-tdp** [comment]
Pull request test result: [success](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/616/) [:star:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/616/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [failure](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/639/) [:x:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/639/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [failure](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/654/) [:x:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/654/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [failure](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/662/) [:x:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/662/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [failure](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/663/) [:x:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/663/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [failure](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/664/) [:x:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/664/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [failure](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/720/) [:x:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/720/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [success](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/740/) [:star:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/740/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [success](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/748/) [:star:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/748/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [success](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/749/) [:star:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/749/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [success](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/753/) [:star:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/753/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [success](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/760/) [:star:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/760/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [success](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/765/) [:star:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/765/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [success](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/766/) [:star:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/766/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [success](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/767/) [:star:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/767/)

File: dataprep-webapp/src/app/components/suggestions-stats/actions-suggestions/_action-suggestions.scss [hunk 1]
```diff
@@ -10,99 +10,78 @@
...
-					border-top: 1px solid transparent;
-					border-right: 1px solid transparent;
-					border-left: 1px solid transparent;
-					box-shadow: none;
-
-					&::-ms-clear {
-						display: none;
-					}
+.actions-suggestions {
+	@include flex-column();
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/suggestions-stats/actions-suggestions/_action-suggestions.scss]: Avoid using Bourbon please

File: dataprep-webapp/src/app/components/suggestions-stats/actions-suggestions/actions-suggestions-directive.spec.js [hunk 2]
```diff
@@ -11,57 +11,45 @@
...
-    var element;
-    var createElement;
-    var stateMock;
-    var body = angular.element('body');
-    beforeEach(angular.mock.module('data-prep.actions-suggestions', function ($provide) {
+    let scope;
+    let element;
+    let createElement;
+    let stateMock;
+    let body = angular.element('body');
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/suggestions-stats/actions-suggestions/actions-suggestions-directive.spec.js]: `const` instead

File: dataprep-webapp/src/app/components/transformation/form/choice/transformation-choice-param-controller.js [hunk 3]
```diff
@@ -39,5 +39,15 @@ export default function TransformChoiceParamCtrl() {
...
 	};
 
+	/**
+	 * @ngdoc method
+	 * @name getLabelByValue
+	 * @methodOf data-prep.transformation-form.controller:TransformChoiceParamCtrl
+	 * @description get label to display by value
+	 */
+	vm.getLabelByValue = (value) => {
+		return _.find(vm.parameter.configuration.values, { value }).label;
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/transformation/form/choice/transformation-choice-param-controller.js]: Avoid using `lodash`

File: dataprep-webapp/src/app/components/transformation/form/choice/transformation-choice-param-directive.js [hunk 4]
```diff
@@ -29,33 +29,36 @@ export default function TransformChoiceParam($rootScope, $compile) {
...
+				.filter(function (optionValue) {
+					return optionValue.parameters && optionValue.parameters.length;
+				})
+				.forEach(function (optionValue) {
+					const isolatedScope = $rootScope.$new(true);
+					isolatedScope.parameter = ctrl.parameter;
+					isolatedScope.optionValue = optionValue;
+					isolatedScope.isReadonly = ctrl.isReadonly;
+
+					const template = '<transform-params ' +
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/transformation/form/choice/transformation-choice-param-directive.js]: You could use ES6 templating feature instead

File: dataprep-webapp/src/app/components/transformation/form/choice/transformation-choice-param-directive.js [hunk 5]
```diff
@@ -29,33 +29,36 @@ export default function TransformChoiceParam($rootScope, $compile) {
...
-	$compile(template)(isolatedScope, function (cloned) {
-		iElement.append(cloned);
-	});
-
-	scope.$on('$destroy', () => {
-		isolatedScope.$destroy();
-	});
-})
-                .value();
+				.filter(function (optionValue) {
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/transformation/form/choice/transformation-choice-param-directive.js]: You could use arrow fn

File: dataprep-webapp/src/app/components/transformation/form/simple/transformation-simple-param.html [hunk 6]
```diff
@@ -14,31 +14,54 @@
   -->
 
 <div class="param">
-    <div class="param-row">
-
+    <div ng-class="{'param-row' : simpleParamCtrl.parameter.type === 'boolean'}">
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/transformation/form/simple/transformation-simple-param.html]: This does not seem to be really template friendly, and it's repeated several times, can we extracted that condition into a function inside the controller?

File: dataprep-webapp/src/app/components/transformation/form/transformation-form-directive.spec.js [hunk 7]
```diff
@@ -11,32 +11,32 @@
...
     }));
 
-    beforeEach(inject(function ($rootScope, $compile) {
+    beforeEach(inject(($rootScope, $compile) => {
         scope = $rootScope.$new();
 
-        createElement = function () {
-            var element = angular.element('<transform-form transformation="transformation" on-submit="onSubmit"></transform-form>');
+        createElement = () => {
+            let element = angular.element('<transform-form transformation="transformation" on-submit="onSubmit" is-readonly="isReadonly"></transform-form>');
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/transformation/form/transformation-form-directive.spec.js]: `const` here

File: dataprep-webapp/src/app/components/transformation/form/transformation-form-directive.spec.js [hunk 8]
```diff
@@ -60,7 +60,7 @@ describe('Transformation params directive', function () {
         };
 
         //when
-        var element = createElement();
+        let element = createElement();
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/transformation/form/transformation-form-directive.spec.js]: `const` here

File: dataprep-webapp/src/app/components/transformation/form/transformation-form-directive.spec.js [hunk 9]
```diff
@@ -93,10 +93,10 @@ describe('Transformation params directive', function () {
...
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/transformation/form/transformation-form-directive.spec.js]: `const` here

File: dataprep-webapp/src/app/components/transformation/form/transformation-form-directive.spec.js [hunk 10]
```diff
@@ -93,10 +93,10 @@ describe('Transformation params directive', function () {
...
 
         //then
-        var paramChoice = element.find('.param').eq(0);
+        let paramChoice = element.find('.param').eq(0);
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/transformation/form/transformation-form-directive.spec.js]: `const` here

File: dataprep-webapp/src/app/components/transformation/form/transformation-form-directive.spec.js [hunk 11]
```diff
@@ -106,9 +106,9 @@ describe('Transformation params directive', function () {
         expect(paramChoice.find('.param-input').eq(0).find('option').eq(1).text()).toBe('noParamChoice2');
     });
 
-    it('should render an action with choice containing parameters', function () {
+    it('should render an action with choice containing parameters', () => {
         //given
-        var parameter = {
+        let parameter = {
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/transformation/form/transformation-form-directive.spec.js]: `const` here

File: dataprep-webapp/src/app/components/transformation/form/transformation-form-directive.spec.js [hunk 12]
```diff
@@ -141,8 +141,8 @@ describe('Transformation params directive', function () {
             name: 'menu with param',
             parameters: [parameter],
         };
-        var element = createElement();
-        var renderedParams = element.find('transform-params').eq(0);
+        let element = createElement();
+        let renderedParams = element.find('transform-params').eq(0);
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/transformation/form/transformation-form-directive.spec.js]: `const` here

File: dataprep-webapp/src/app/components/transformation/form/transformation-form-directive.spec.js [hunk 13]
```diff
@@ -141,8 +141,8 @@ describe('Transformation params directive', function () {
             name: 'menu with param',
             parameters: [parameter],
         };
-        var element = createElement();
-        var renderedParams = element.find('transform-params').eq(0);
+        let element = createElement();
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/transformation/form/transformation-form-directive.spec.js]: `const` here

File: dataprep-webapp/src/app/components/transformation/form/transformation-form-directive.spec.js [hunk 14]
```diff
@@ -258,13 +258,13 @@ describe('Transformation params directive', function () {
         };
 
         //when
-        var element = createElement();
+        let element = createElement();
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/transformation/form/transformation-form-directive.spec.js]: `const` here too

File: dataprep-webapp/src/app/components/widgets/editable-regex/editable-regex.html [hunk 15]
```diff
@@ -13,21 +13,36 @@
...
-    </talend-dropdown>
+<div class="editable-regex" ng-switch="!!editableRegexCtrl.isReadonly">
+    <div ng-switch-when="false">
+        <talend-dropdown force-side="left">
+            <button class="dropdown-action" type="button">{{editableRegexCtrl.getTypeKey()}}</button>
+            <ul class="dropdown-menu">
+                <li ng-repeat="type in ::editableRegexCtrl.types track by $index"
+                    ng-click="editableRegexCtrl.setSelectedType(type)"
+                    class="regex-type-item">
+                    <div class="regex-type-item-key">{{::type.key}}</div>
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/widgets/editable-regex/editable-regex.html]: `span` here (I know is already there before)

File: dataprep-webapp/src/app/components/widgets/editable-regex/editable-regex.html [hunk 16]
```diff
@@ -13,21 +13,36 @@
...
+                    <div class="regex-type-item-label">{{::type.label}}</div>
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/widgets/editable-regex/editable-regex.html]: `span` here

File: dataprep-webapp/src/app/components/suggestions-stats/actions-suggestions/_action-suggestions.scss (hunk 1 above)
Comment (charlesnguyen) [inline dataprep-webapp/src/app/components/suggestions-stats/actions-suggestions/_action-suggestions.scss]: I think it is not from Bourbon. It is a TDP mixin (_container.scss)

File: dataprep-webapp/src/app/components/suggestions-stats/actions-suggestions/actions-suggestions-directive.spec.js (hunk 2 above)
Comment (charlesnguyen) [inline dataprep-webapp/src/app/components/suggestions-stats/actions-suggestions/actions-suggestions-directive.spec.js]: done

File: dataprep-webapp/src/app/components/transformation/form/choice/transformation-choice-param-controller.js (hunk 3 above)
Comment (charlesnguyen) [inline dataprep-webapp/src/app/components/transformation/form/choice/transformation-choice-param-controller.js]: done

File: dataprep-webapp/src/app/components/transformation/form/choice/transformation-choice-param-directive.js (hunk 5 above)
Comment (charlesnguyen) [inline dataprep-webapp/src/app/components/transformation/form/choice/transformation-choice-param-directive.js]: done

File: dataprep-webapp/src/app/components/transformation/form/choice/transformation-choice-param-directive.js (hunk 4 above)
Comment (charlesnguyen) [inline dataprep-webapp/src/app/components/transformation/form/choice/transformation-choice-param-directive.js]: done

File: dataprep-webapp/src/app/components/transformation/form/transformation-form-directive.spec.js (hunk 7 above)
Comment (charlesnguyen) [inline dataprep-webapp/src/app/components/transformation/form/transformation-form-directive.spec.js]: done

File: dataprep-webapp/src/app/components/transformation/form/transformation-form-directive.spec.js (hunk 8 above)
Comment (charlesnguyen) [inline dataprep-webapp/src/app/components/transformation/form/transformation-form-directive.spec.js]: done

File: dataprep-webapp/src/app/components/widgets/editable-regex/editable-regex.html (hunk 15 above)
Comment (charlesnguyen) [inline dataprep-webapp/src/app/components/widgets/editable-regex/editable-regex.html]: done

File: dataprep-webapp/src/app/components/widgets/editable-regex/editable-regex.html (hunk 16 above)
Comment (charlesnguyen) [inline dataprep-webapp/src/app/components/widgets/editable-regex/editable-regex.html]: done

File: dataprep-webapp/src/app/components/transformation/form/simple/transformation-simple-param.html (hunk 6 above)
Comment (charlesnguyen) [inline dataprep-webapp/src/app/components/transformation/form/simple/transformation-simple-param.html]: done

**Comment by build-talend-tdp** [comment]
Pull request test result: [success](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/803/) [:star:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/803/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [success](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/804/) [:star:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/804/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [failure](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/814/) [:x:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/814/)

**Pull request merged**